pinecone/integration_test_suite.go

```diff
@@ -0,0 +1,218 @@
+	clientSourceTag   Client
+	idxConnSourceTag  *IndexConnection
+}
+
+func (ts *IntegrationTests) SetupSuite() {
+	ctx := context.Background()
+
+	additionalMetadata := map[string]string{"api-key": ts.apiKey}
+
+	namespace, err := uuid.NewUUID()
+	require.NoError(ts.T(), err)
+
+	idxConn, err := newIndexConnection(newIndexParameters{
+		additionalMetadata: additionalMetadata,
+		host:               ts.host,
+		namespace:          namespace.String(),
+		sourceTag:          ""})
+	require.NoError(ts.T(), err)
+	require.NotNil(ts.T(), idxConn, "Failed to create idxConn")
+
+	ts.idxConn = idxConn
+
+	// Deterministically create vectors
+	vectors := createVectorsForUpsert()
+
+	// Set vector IDs
+	vectorIds := make([]string, len(vectors))
+	for i, v := range vectors {
+		vectorIds[i] = v.Id
+	}
+	ts.vectorIds = vectorIds
+
+	// Upsert vectors
+	err = upsertVectors(ts, ctx, vectors)
+	if err != nil {
+		log.Fatalf("Failed to upsert vectors in SetupSuite: %v", err)
+	}
+
+	ts.sourceTag = "test_source_tag"
+	idxConnSourceTag, err := newIndexConnection(newIndexParameters{
+		additionalMetadata: additionalMetadata,
+		host:               ts.host,
+		namespace:          namespace.String(),
+		sourceTag:          ts.sourceTag})
+	require.NoError(ts.T(), err)
+	ts.idxConnSourceTag = idxConnSourceTag
+
+	fmt.Printf("\n %s set up suite completed successfully\n", ts.indexType)
+}
+
+func (ts *IntegrationTests) TearDownSuite() {
+	ctx := context.Background()
+
+	// Delete test indexes
+	err := ts.client.DeleteIndex(ctx, ts.serverlessIdxName)
+	err = ts.client.DeleteIndex(ctx, ts.podIdxName)
+
+	err = ts.idxConn.Close()
+	require.NoError(ts.T(), err)
+
+	err = ts.idxConnSourceTag.Close()
+	require.NoError(ts.T(), err)
+	fmt.Printf("\n %s setup suite torn down successfully\n", ts.indexType)
+}
+
+// Helper funcs
+func GenerateTestIndexName() string {
+	return fmt.Sprintf("index-%d", time.Now().UnixMilli())
+}
+
+func upsertVectors(ts *IntegrationTests, ctx context.Context, vectors []*Vector) error {
+	maxRetries := 12
+	delay := 12 * time.Second
+	fmt.Printf("Attempting to upsert vectors into host \"%s\"...\n", ts.host)
+	for i := 0; i < maxRetries; i++ {
+		ready, err := GetIndexStatus(ts, ctx)
+		if err != nil {
+			fmt.Printf("Error getting index ready: %v\n", err)
+			return err
+		}
+		if ready {
+			upsertVectors, err := ts.idxConn.UpsertVectors(ctx, vectors)
+			require.NoError(ts.T(), err)
+			fmt.Printf("Upserted vectors: %v into host: %s\n", upsertVectors, ts.host)
+			break
+		} else {
+			time.Sleep(delay)
+			fmt.Printf("Host \"%s\" not ready for upserting yet, retrying... (%d/%d)\n", ts.host, i, maxRetries)
+		}
+	}
+	return nil
+}
+
+// TODO: how to get this func to work for client tests too
+func GetIndexStatus(ts *IntegrationTests, ctx context.Context) (bool, error) {
+	var indexName string
+	if ts.indexType == "serverless" {
+		indexName = ts.serverlessIdxName
+	} else if ts.indexType == "pods" {
+		indexName = ts.podIdxName
+	}
```

Comment:
I feel like needing to do these `indexType` checks inside of this method is harder to reason about, couldn't we just take in an `indexName` as an argument and then make this function handle one index no matter what?

Reply:
gone!